File: src/input/core/turbo.c
#include <assert.h>
#include <stddef.h>   /* NULL */
#include "turbo.h"

/*
 * A faithful port of ../../input/turbo.py -- the per-button `_since` dict becomes a
 * fixed since[] table gated by a `tracking` mask (a bit is tracked iff it has been
 * held continuously and thus has a valid press tick). Keep the two in lockstep;
 * input/dst.py tier I11 proves they agree under fuzzing.
 */

void turbo_init(turbo_t *t, buttons_t buttons, uint32_t on_ticks, uint32_t off_ticks)
{
    assert(t != NULL);
    assert(on_ticks >= 1u && on_ticks <= TURBO_MAX_TICKS);
    assert(off_ticks >= 1u && off_ticks <= TURBO_MAX_TICKS);
    t->buttons = buttons;
    t->on_ticks = on_ticks;
    t->off_ticks = off_ticks;
    t->period = on_ticks + off_ticks;
    t->tracking = 0;
    t->started = false;
    t->last_now = 0;
    for (uint32_t i = 0; i < BUTTONS_BITS; i++)
        t->since[i] = 0;
}

buttons_t turbo_step(turbo_t *t, uint32_t now, buttons_t pressed)
{
    assert(t != NULL);
    assert(!t->started || now >= t->last_now);   /* time must not go backwards */

    buttons_t out = pressed;
    for (uint32_t i = 0; i < BUTTONS_BITS; i++) {
        buttons_t bit = (buttons_t)1u << i;
        if (!(t->buttons & bit))
            continue;                             /* not a turbo button */
        if (pressed & bit) {
            if (!(t->tracking & bit)) {           /* rising edge -> start ON this tick */
                t->tracking |= bit;
                t->since[i] = now;
            }
            assert(t->since[i] <= now);
            if ((now - t->since[i]) % t->period >= t->on_ticks)
                out &= ~bit;                      /* OFF portion of the pulse */
        } else {
            t->tracking &= ~bit;                  /* released -> forget phase */
        }
    }

    t->started = true;
    t->last_now = now;
    /* negative space: turbo only gates presses off, it never invents one */
    assert((out & ~pressed) == 0);
    return out;
}

uint32_t turbo_next_toggle(const turbo_t *t, uint32_t now)
{
    assert(t != NULL);
    assert(t->period >= 2u);                          /* on+off, each >= 1 */
    uint32_t best = 0u;                               /* 0 = no held turbo button -> nothing due */
    for (uint32_t i = 0; i < BUTTONS_BITS; i++) {
        buttons_t bit = (buttons_t)1u << i;
        if (!(t->tracking & bit))
            continue;                                 /* only a held turbo button toggles */
        assert(t->since[i] <= now);
        uint32_t phase = (now - t->since[i]) % t->period;
        /* ON while phase < on_ticks, OFF otherwise; the next toggle is the next boundary. */
        uint32_t until = (phase < t->on_ticks) ? (t->on_ticks - phase) : (t->period - phase);
        assert(until >= 1u && until <= t->period);    /* a toggle is always strictly ahead */
        if (best == 0u || until < best)
            best = until;
    }
    return best;
}

File: test/reference/input/turbo.py
"""
Turbo (auto-fire) — a post-SOCD output-modulation stage.

While a turbo-enabled button is physically held, its OUTPUT is pulsed on/off at a
fixed rate, so one hold becomes repeated presses (auto-fire). This runs AFTER SOCD
(raw → … → SOCD → turbo → buffer): turbo modulates the already-resolved button,
exactly where GP2040-CE loads it ("close to the end").

The pulse is derived from ABSOLUTE time, not an incrementing counter: each held
button remembers the tick it was pressed, and its phase is `(now - pressed_at) %
period`. It is ON for the first `on_ticks` of every period and OFF for the rest.
Two consequences that matter for feel and for testing:
  * the rising edge is always ON (phase 0 < on_ticks) — the first press registers
    immediately, never a dropped first frame;
  * a jump in `now` (the sync window can advance time by more than one tick) still
    lands on the correct phase, because phase is absolute, not accumulated.

Turbo can only ever GATE a press off; it never fabricates one (out ⊆ in). Buttons
not in the turbo set pass through untouched.

Engineering standard: static per-button state, bounded (one compare per button),
asserts on the range + the never-fabricate negative space, named limits.
"""
from dataclasses import dataclass

MAX_TURBO_TICKS = 255           # per-phase cap; fits a u8 counter on-chip


@dataclass
class TurboConfig:
    buttons: frozenset = frozenset()   # which button names auto-fire
    on_ticks: int = 1                  # ON portion of each pulse period
    off_ticks: int = 1                 # OFF portion of each pulse period

    def __post_init__(self):
        self.buttons = frozenset(self.buttons)
        assert 1 <= self.on_ticks <= MAX_TURBO_TICKS, "on_ticks out of range"
        assert 1 <= self.off_ticks <= MAX_TURBO_TICKS, "off_ticks out of range"


class Turbo:
    def __init__(self, cfg=None):
        self.cfg = cfg or TurboConfig()
        self._period = self.cfg.on_ticks + self.cfg.off_ticks
        self._since = {}            # button -> tick it was pressed (only while held)
        self._now = -1

    def step(self, now, pressed):
        """Advance to `now` with the post-SOCD pressed set; return the gated set."""
        assert now >= self._now, "time must not go backwards"
        pressed = set(pressed)
        out = set(pressed)
        for b in self.cfg.buttons:
            if b in pressed:
                since = self._since.get(b)
                if since is None:               # rising edge -> start ON this tick
                    since = now
                    self._since[b] = now
                assert since <= now, "press tick is in the future"
                if (now - since) % self._period >= self.cfg.on_ticks:
                    out.discard(b)              # OFF portion of the pulse
            else:
                self._since.pop(b, None)        # released -> forget phase
        self._now = now
        # negative space: turbo only gates presses off, it never invents one.
        assert out <= pressed, "turbo fabricated a press"
        return out
